recirq/quantum_chess/circuit_transformer.py

```diff
@@ -305,6 +317,384 @@ def transform(self, circuit: cirq.Circuit) -> cirq.Circuit:
+        self,
+        g: Dict[cirq.GridQubit, List[cirq.GridQubit]],
+        s: cirq.GridQubit,
+        t: cirq.GridQubit,
+    ) -> int:
+        """Returns the shortest distance between the source and target qubits.
+
+        Uses breadth-first search traversal.
+
+        Args:
+          g: A physical qubits graph.
+          s: The source qubit from which to start breadth-first search.
+          t: The target qubit to search.
+        """
+        dist = defaultdict(int)
+        visited = set()
+        visited.add(s)
+        queue = deque()
+        queue.append(s)
+        while queue:
+            q = queue.popleft()
+            if q == t:
+                return dist[t]
+            for neighbor in g[q]:
+                if neighbor not in visited:
+                    dist[neighbor] = dist[q] + 1
+                    visited.add(neighbor)
+                    queue.append(neighbor)
+        return math.inf
+
+    def calculate_initial_mapping(
+        self,
+        pg: Dict[cirq.Qid, List[cirq.Qid]],
+        lg: Dict[cirq.Qid, List[Tuple[cirq.Qid, int]]],
+    ) -> Dict[cirq.Qid, cirq.GridQubit]:
+        """Returns an initial mapping of logical qubits to physical qubits.
+
+        This initial mapping algorithm is proposed by the paper "A Dynamic
+        Look-Ahead Heuristic for the Qubit Mapping Problem of NISQ Computer:
+        https://ieeexplore.ieee.org/abstract/document/8976109.
+
+        Args:
+          pg: A physical qubits graph.
+          lg: A logical qubits graph.
+        """
+        mapping = defaultdict()
+
+        pg_center = self.find_graph_center(pg)
+        lg_center = self.find_graph_center(lg)
+        mapping[lg_center] = pg_center
+
+        traversal_order = self.traverse(lg, lg_center)
+
+        while traversal_order:
+            lq = traversal_order.popleft()
+            if lq == lg_center:
+                continue
+            pq = None
+            reference_qubits = self.find_reference_qubits(mapping, lg, lq)
+            ref_q = reference_qubits[0]
+            candidate_qubits = self.find_candidate_qubits(mapping.values(), pg, ref_q)
+            if len(reference_qubits) > 1:
+                # For each reference location, find the shortest path distance
+                # to each of the candidate qubits. Only keep the nearest
+                # candidate qubits with smallest distance.
+                for ref_q in reference_qubits[1:]:
+                    distances = defaultdict(list)
+                    for cand_q in candidate_qubits:
+                        d = self.find_shortest_path_distance(pg, ref_q, cand_q)
+                        distances[d].append(cand_q)
+                    nearest_candidate_qubits = None
+                    min_dist = math.inf
+                    for dist in distances:
+                        if dist < min_dist:
+                            min_dist = dist
+                            nearest_candidate_qubits = distances[dist]
```

Comment:
nit: could we replace this block with a min() call? It looks like
```
min_dist = min(distances.keys())
candidate_qubits = distances[min_dist]
```

Reply:
done!